### ISSUE TITLE
Flash Attention v2 for Roberta-large

### ISSUE THREAD

**Issue by jamesharrisivi** [huggingface/optimum#1711]
### Feature request

```
from optimum.bettertransformer import BetterTransformer
self.model = BetterTransformer.transform(self.model)
```
Does provide 2x speed up compared to torch, but it would be great to integrate further speed up and use https://github.com/iliaschalkidis/flash-roberta to have flash attention v2 and torch compilation (torch 2).

### Motivation

Roberta-large further optimizations for inference.

### Your contribution

I can help run inference tests.

**Comment by fxmarty** [comment]
@jamesharrisivi SDPA (that can dispatch to FA2) support for BERT is being merged natively in Transformers: https://github.com/huggingface/transformers/pull/28802

This would likely make it easy to use SDPA as well for Roberta natively there.

Going forward, we are privileging direct support of SDPA/torch compile in Transformers and will not be adding new features to BetterTransformer. You can read more here: https://huggingface.co/docs/transformers/main/en/perf_infer_gpu_one

**Comment by jamesharrisivi** [comment]
Ok thank you.

I see the current implementation uses https://github.com/pytorch/pytorch/blob/0f5e24bda9450a89ba56d2fdd471f56d97fe4546/aten/src/ATen/native/transformers/transformer.cpp#L75

But this doesn't use flash attention or does it?

**Comment by fxmarty** [comment]
It does, it uses https://github.com/pytorch/pytorch/blob/0f5e24bda9450a89ba56d2fdd471f56d97fe4546/aten/src/ATen/native/transformers/cuda/attention.cu#L467, see https://github.com/pytorch/pytorch/blob/232f09e0ea7b7d9ac9752bd366aa2f4c372fd9d9/aten/src/ATen/native/native_functions.yaml#L14505

This one is able to dispatch to flash attention if the attn_mask argument is None (`mask` in the C++ code)

**Comment by fxmarty** [comment]
@michaelfeil Yes, masking works for training. We simply drop the mask when the `is_causal` argument of SDPA can be used: https://github.com/huggingface/transformers/blob/83e366bfd49708796e2c6461d3988d23d008502a/src/transformers/modeling_attn_mask_utils.py#L375-L380
